cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringAgentService.java

```diff
@@ -123,20 +175,24 @@ protected long runTask() {
         .collect(Collectors.toList());
     } catch (IOException e) {
       LOG.warn("Failed to get peer information", e);
-      return connectionInterval;
+      TimeUnit.MILLISECONDS.sleep(connectionInterval);
+      return;
     }
     for (PeerInfo peer : peers) {
       try {
         Preconditions.checkArgument(peer.getEndpoint() != null,
                                     "Peer %s doesn't have an endpoint", peer.getName());
         String uri = CONNECT_CONTROL_CHANNEL + instanceName;
-        String lastMessageId = lastMessageIds.get(peer);
-        if (lastMessageId != null) {
-          uri = uri + "?messageId=" + URLEncoder.encode(lastMessageId, "UTF-8");
+        String lastControlMessageId = peerToLastControlMessageIds.get(peer);
+        if (lastControlMessageId != null) {
+          uri = uri + "?messageId=" + URLEncoder.encode(lastControlMessageId, "UTF-8");
```

Comment:
we shouldn't send parameters with a mix of query params and request body. Should create some request class that contains the lastControlMessageId and a list of Notifications.

Reply:
fixed